dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/pingcap/tiflow/dm/pkg/log"
+)
+
+type connAmountChecker struct {
+	targetDB *conn.BaseDB
+	stCfgs   []*config.SubTaskConfig
+
+	getConfigConn func(stCfgs []*config.SubTaskConfig) int
+	workerName    string
+}
+
+func newConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig, fn func(stCfgs []*config.SubTaskConfig) int, workerName string) connAmountChecker {
+	return connAmountChecker{
+		targetDB:      targetDB,
+		stCfgs:        stCfgs,
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified amount of connection exceeds downstream's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.targetDB.GetBaseConn(ctx)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW GLOBAL VARIABLES LIKE 'max_connections'")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer rows.Close()
+	var (
+		maxConn   int
+		_variable string
+	)
+	for rows.Next() {
+		err = rows.Scan(&_variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	neededConn := c.getConfigConn(c.stCfgs)
+	if neededConn > maxConn {
+		result.Errors = append(
+			result.Errors,
+			NewError("downstream's max_connections: %d is less than the amount %s needs: %d", maxConn, c.workerName, neededConn),
+		)
+		result.Instruction = "set larger max_connections or reduce the pool size of dm"
+		result.State = StateFailure
+	} else {
+		result.State = StateSuccess
+	}
+	return result
+}
+
+type SyncerConnAmountChecker struct {
+	connAmountChecker
+}
+
+func NewSyncerConnAmountCheker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &SyncerConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			syncerConn := 0
+			for _, stCfg := range stCfgs {
+				// 1. worker count
+				// 2. checkpoint
+				// 3. ddl connection
+				// 4. downstream tracker
+				// 5. shard group keeper
+				// 6. online ddl
+				syncerConn += stCfg.SyncerConfig.WorkerCount + 5
+			}
+			return syncerConn
+		}, "syncer"),
+	}
+}
+
+func (s *SyncerConnAmountChecker) Name() string {
+	return "syncer_conn_amount_checker"
+}
+
+func (s *SyncerConnAmountChecker) Check(ctx context.Context) *Result {
+	return s.check(ctx, s.Name())
+}
+
+type LoaderConnAmountChecker struct {
+	connAmountChecker
+}
+
+func NewLoaderConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &LoaderConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			loaderConn := 0
+			for _, stCfg := range stCfgs {
+				// loader's worker and checkpoint (always keeps one db connection)
```

Comment:
did you check lightning loader?

Reply:
It seems only one db connection is used for restoring in Lightning (another is for checkpoint). What about skipping this check when using lightning (not really necessary in this case)?

FYI: https://github.com/pingcap/tidb/blob/0e19c1cd647a63fbbabd253fe18a0d54478fee41/br/pkg/lightning/backend/tidb/tidb.go#L94

Reply:
It seems loader's config doesn't affect lightning's concurrency, which is related to concurrency of the physical machine.

Reply:
we should be able to change its concurrency, I'll add a working item to https://github.com/pingcap/tiflow/issues/3510 .

And lightning will become the default load unit implementation, so we should not skip checking it. Will think about it later

Reply:
Now I think we should assign the concurrency of tidb backend (cfg.App.RegionConcurrency ?) to LoaderConfig.PoolSize. We can finish it first before this PR, or you left a TODO in this PR and we merge this PR first.

Reply:
Have you filed a PR to resolve it? Perhaps I can leave a TODO here and see if I could work it out (in another PR to solve this).

Reply:
We can left a TODO here